// @ls1intum/apollon/lib/es5/packages/uml-deployment-diagram/uml-deployment-node/uml-deployment-node-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UMLDeploymentNodeUpdate = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var react_redux_1 = require("react-redux");
var button_1 = require("../../../components/controls/button/button");
var color_button_1 = require("../../../components/controls/color-button/color-button");
var divider_1 = require("../../../components/controls/divider/divider");
var trash_1 = require("../../../components/controls/icon/trash");
var textfield_1 = require("../../../components/controls/textfield/textfield");
var style_pane_1 = require("../../../components/style-pane/style-pane");
var styles_1 = require("../../../components/theme/styles");
var uml_element_repository_1 = require("../../../services/uml-element/uml-element-repository");
var Flex = styles_1.styled.div(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  display: flex;\n  align-items: baseline;\n  justify-content: space-between;\n"], ["\n  display: flex;\n  align-items: baseline;\n  justify-content: space-between;\n"])));
var DeploymentNodeUpdate = /** @class */ (function (_super) {
    tslib_1.__extends(DeploymentNodeUpdate, _super);
    function DeploymentNodeUpdate() {
        var _this = _super.apply(this, tslib_1.__spreadArray([], tslib_1.__read(arguments), false)) || this;
        _this.state = { colorOpen: false };
        _this.toggleColor = function () {
            _this.setState(function (state) { return ({
                colorOpen: !state.colorOpen,
            }); });
        };
        _this.onRename = function (value) {
            var _a = _this.props, element = _a.element, update = _a.update;
            update(element.id, { name: value });
        };
        _this.onUpdate = function (value) {
            var _a = _this.props, element = _a.element, update = _a.update;
            update(element.id, { stereotype: value });
        };
        return _this;
    }
    DeploymentNodeUpdate.prototype.render = function () {
        var _this = this;
        var element = this.props.element;
        return (react_1.default.createElement("div", null,
            react_1.default.createElement("section", null,
                react_1.default.createElement(Flex, null,
                    react_1.default.createElement(textfield_1.Textfield, { value: element.name, onChange: this.onRename, autoFocus: true }),
                    react_1.default.createElement(color_button_1.ColorButton, { onClick: this.toggleColor }),
                    react_1.default.createElement(button_1.Button, { color: "link", tabIndex: -1, onClick: function () { return _this.props.delete(element.id); } },
                        react_1.default.createElement(trash_1.TrashIcon, null))),
                react_1.default.createElement(style_pane_1.StylePane, { open: this.state.colorOpen, element: element, onColorChange: this.props.update, lineColor: true, textColor: true, fillColor: true })),
            react_1.default.createElement("section", null,
                react_1.default.createElement(divider_1.Divider, null),
                react_1.default.createElement(Flex, null,
                    react_1.default.createElement(textfield_1.Textfield, { value: element.stereotype, onChange: this.onUpdate })))));
    };
    return DeploymentNodeUpdate;
}(react_1.Component));
var enhance = (0, react_redux_1.connect)(null, {
    update: uml_element_repository_1.UMLElementRepository.update,
    delete: uml_element_repository_1.UMLElementRepository.delete,
});
exports.UMLDeploymentNodeUpdate = enhance(DeploymentNodeUpdate);
var templateObject_1;
