// @ls1intum/apollon/lib/es5/components/create-pane/preview-element-component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreviewElementComponent = exports.Preview = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importStar(require("react"));
var draggable_1 = require("../draggable/draggable");
var styles_1 = require("../theme/styles");
var canvas_element_1 = require("../uml-element/canvas-element");
var hoverable_1 = require("../uml-element/hoverable/hoverable");
exports.Preview = (0, styles_1.styled)((0, hoverable_1.hoverable)(canvas_element_1.CanvasElement)).attrs(function (props) { return ({
    child: canvas_element_1.CanvasElement,
    scale: props.scale,
}); })(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n  overflow: visible;\n  fill: white;\n  scale: ", ";\n  transform-origin: center center;\n"], ["\n  overflow: visible;\n  fill: white;\n  scale: ", ";\n  transform-origin: center center;\n"])), function (props) { var _a; return (_a = props.scale) !== null && _a !== void 0 ? _a : 0.8; });
var PreviewElementComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PreviewElementComponent, _super);
    function PreviewElementComponent() {
        var _this = _super.apply(this, tslib_1.__spreadArray([], tslib_1.__read(arguments), false)) || this;
        _this.onDrop = function (event) {
            var element = _this.props.element.clone({
                bounds: tslib_1.__assign(tslib_1.__assign({}, _this.props.element.bounds), event.position),
            });
            _this.props.create(element, event.owner);
        };
        return _this;
    }
    PreviewElementComponent.prototype.render = function () {
        var element = this.props.element;
        return (react_1.default.createElement(draggable_1.Draggable, { onDrop: this.onDrop },
            react_1.default.createElement(exports.Preview, { id: element.id })));
    };
    return PreviewElementComponent;
}(react_1.Component));
exports.PreviewElementComponent = PreviewElementComponent;
var templateObject_1;
